### PR TITLE
Consistently present absent stdio handles on Windows as NULL handles.

### PR DESCRIPTION
**Pull request by sunfishcode**
This addresses #90964 by making the std API consistent about presenting
absent stdio handles on Windows as NULL handles. Stdio handles may be
absent due to `#![windows_subsystem = "windows"]`, due to the console
being detached, or due to a child process having been launched from a
parent where stdio handles are absent.

Specifically, this fixes the case of child processes of parents with absent
stdio, which previously ended up with `stdin().as_raw_handle()` returning
`INVALID_HANDLE_VALUE`, which was surprising, and which overlapped with an
unrelated valid handle value. With this patch, `stdin().as_raw_handle()`
now returns null in these situation, which is consistent with what it
does in the parent process.

And, document this in the "Windows Portability Considerations" sections of
the relevant documentation.

### CODE REVIEW

**Comment by joshtriplett** [comment]
What do you mean by "which is consistent with what it does in the parent process"? It sounds like GetStdHandle returns `INVALID_HANDLE_VALUE` for this case.

**Comment by sunfishcode** [comment]
> What do you mean by "which is consistent with what it does in the parent process"? It sounds like GetStdHandle returns `INVALID_HANDLE_VALUE` for this case.

In a parent with a detached console, its `GetStdHandle` returns NULL, and Rust's `Stdin::as_raw_handle()` returns NULL.

Currently, in the child of such a parent, `Stdin::as_raw_handle()` returns `INVALID_HANDLE_VALUE`. With this patch, it will return NULL, the same as `Stdin::as_raw_handle` does in the parent process.

**Comment by dtolnay** [comment]
@rust-lang/libs:
@rfcbot fcp merge

This PR concerns the behavior of the stable [AsRawHandle](https://doc.rust-lang.org/std/os/windows/io/trait.AsRawHandle.html) on Windows. Previously a process with absent stdio would return null ptr for its stdio handles, **but** child processes of that process would return INVALID_HANDLE_VALUE (which is `!0`) for its stdio handles. This discrepancy is surprising and problematic because INVALID_HANDLE_VALUE aliases the value of the *current process* handle.

I haven't read all of #90964, though hopefully one of my teammates more knowledgeable on Windows will take a closer look. The justification in https://github.com/rust-lang/rust/issues/90964#issuecomment-1020122293 seems compelling to me. Though be aware this solution diverges from [the preference of a set of anonymous Microsoft consultees](https://github.com/rust-lang/rust/issues/90964#issuecomment-1019505844).

**Comment by joshtriplett** [comment]
Discussed some minor tweaks for readability of the repeated handle usage with @sunfishcode (adding a helper function), but otherwise LGTM.

@rfcbot reviewed

**Comment by sunfishcode** [comment]
Looking at the code, it looks like those redundant checks are not actually needed. I've now submitted a patch which removes them.

File: library/std/src/os/windows/io/raw.rs [hunk 1]
```diff
@@ -96,45 +97,57 @@ impl AsRawHandle for fs::File {
...
 
+// Translate a handle returned from `GetStdHandle` into a handle to return to
+// the user.
+fn stdio_handle(raw: RawHandle) -> RawHandle {
+    // `GetStdHandle` isn't expected to actually fail, so when it returns
+    // `INVALID_HANDLE_VALUE`, it means we were launched from a parent which
+    // didn't provide us with stdio handles, such as a parent with a detached
+    // console. In that case, return null to the user, which is consistent
+    // with what they'd get in the parent, and which avoids the problem that
+    // `INVALID_HANDLE_VALUE` aliases the current process handle.
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply library/std/src/os/windows/io/raw.rs:147]: In Windows, the current process handle, as returned by  [`GetCurrentProcess`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentprocess#remarks) has type `HANDLE`, aka `io::RawHandle` here, which *is* the same as the return type of eg. [`CreateFileW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew).

`SOCKET`, aka `io::RawSocket` here, on the other hand, is a different type.

[the remark this answers is not in the document]
Comment (pravic) [reply library/std/src/os/windows/io/raw.rs:147]: And so does a dozens of other types that are also an alias (or a [newtype](https://stackoverflow.com/questions/3429523/what-does-this-line-of-c-c-preprocessor-mean) if compiled with `-DSTRICT`) of `HANDLE`: https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types

So, in fact it should be a newtype in Rust as well.

[the remark this answers is not in the document]
Comment (sunfishcode) [reply library/std/src/os/windows/io/raw.rs:147]: Some of the types linked there using `HANDLE` are so different that they don't even use `CloseHandle`. We already do document that it's not valid to use `OwnedHandle`/`BorrowedHandle` to hold those kinds of handles.

Beyond that though, `OwnedHandle`, `BorrowedHandle`, and `AsHandle` are intended to be used with both file handles and process handles, for three reasons:
 - Rust already uses the same `RawHandle`/`AsRawHandle`/`IntoRawHandle` types and traits for both [processes](https://doc.rust-lang.org/stable/std/process/struct.Child.html#impl-AsRawHandle) and [files](https://doc.rust-lang.org/stable/std/fs/struct.File.html#impl-AsRawHandle), so it's consistent with existing APIs in Rust
 - Both process handles and file handles can be passed to `CloseHandle`, `GetHandleInformation`, `SetHandleInformation`, `DuplicateHandle`, and others, so they may be different, but they're not completely different.
 - Rust also does have distinct higher-level types for different kinds of handles, such as `std::fs::File` for file(-like) handles, and `std::process::Child` for process handles, and these are what most Rust code uses most of the time, so it isn't as important to make a static type distinction at the `OwnedHandle`/etc. level.

**Review by dtolnay — APPROVED** [review]

**Pull request merged**